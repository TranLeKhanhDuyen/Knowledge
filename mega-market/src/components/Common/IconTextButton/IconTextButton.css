.icontext-btn-container {
  display: flex;
  align-items: center;
  border: none;
  background-color: transparent;
  cursor: pointer;
  gap: 6px;
  width: max-content;
}

.icontext-btn-container-right {
  flex-direction: row-reverse;
}

.select.icontext-btn-container {
  background-color: var(--tertiary-color);
  padding: 9px 14px;
  border-radius: 18px;
}

.select .icontext-title {
  color: var(--black-color);
}

.icontext-title {
  color: var(--text-secondary-color);
}

.icontext-title-sm {
  font-size: var(--font-sm);
}

.icontext-title-md {
  font-size: var(--font-md);
}

.icontext-sub-title {
  font-weight: var(--fw-bold);
}

.img-icontext-btn-sm {
  width: 18px;
  height: 18px;
}

.img-icontext-btn-md {
  width: 24px;
  height: 24px;
}
