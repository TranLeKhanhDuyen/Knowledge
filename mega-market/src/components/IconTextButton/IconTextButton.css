.icontext-btn-container {
  display: flex;
  align-items: center;
  border: none;
  background-color: transparent;
  cursor: pointer;
  gap: 6px;
}

.icontext-btn-container-right {
  flex-direction: row-reverse;
}

.icontext-title {
  color: var(--text-secondary-color);
}

.select .icontext-title {
  color: var(--black-color);
}

.icontext-title-sm {
  font-size: var(--font-sm);
}

.icontext-title-md {
  font-size: var(--font-md);
}

.icontext-sub-title {
  font-weight: var(--fw-bold);
}

.img-icontext-btn-sm {
  width: 18px;
  height: 18px;
}

.img-icontext-btn-md {
  width: 24px;
  height: 24px;
}
