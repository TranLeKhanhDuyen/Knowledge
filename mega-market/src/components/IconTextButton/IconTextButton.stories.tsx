import type { Meta, StoryObj } from '@storybook/react'
import locationIcon from '@assets/images/icons/location.svg'
import truckIcon from '@assets/images/icons/iconoir-delivery-truck.svg'
import arrowDownIcon from '@assets/images/icons/arrow-down-1.svg'

// Components
import { IconTextButton } from '@components'

export default {
  title: 'Components/IconTextButton',
  component: IconTextButton,
  tags: ['autodocs'],
  argTypes: {
    size: { description: 'The size of button' },
    additionalClass: { description: 'Add class to the button' },
    title: { description: 'The title of button.' },
    subTitle: { description: 'The subs title of button.' },
    iconPosition: {
      description: 'The position of the icon relative to the text',
      control: {
        type: 'radio',
        options: ['left', 'right']
      }
    }
  }
} as Meta

type Story = StoryObj<typeof IconTextButton>

export const Default: Story = {
  args: {
    size: 'sm',
    icon: locationIcon,
    title: 'Deliver to',
    subTitle: '423651'
  }
}

export const IconLeft: Story = {
  ...Default,
  args: {
    icon: truckIcon,
    title: 'Track your order',
    iconPosition: 'left'
  }
}

export const IconRight: Story = {
  ...Default,
  args: {
    icon: arrowDownIcon,
    title: 'Groceries',
    iconPosition: 'right',
    additionalClass: 'select'
  }
}
