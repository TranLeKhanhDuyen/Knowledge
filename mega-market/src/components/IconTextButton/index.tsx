import '@components/IconTextButton/IconTextButton.css'

type TIconTextButtonSize = 'sm' | 'md' | 'lg'
type TIconPosition = 'left' | 'right'

export interface IIconTextButtonProps
  extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  size?: TIconTextButtonSize
  iconPosition?: TIconPosition
  title?: string
  subTitle?: string
  icon?: string
  additionalClass?: string
}

const IconTextButton: React.FC<IIconTextButtonProps> = ({
  size = 'sm',
  iconPosition = 'left',
  title,
  subTitle,
  icon,
  additionalClass,
  ...props
}) => {
  const containerClass = `icontext-btn-container icontext-btn-container-${size} ${
    iconPosition === 'right' ? 'icontext-btn-container-right' : ''
  } ${additionalClass ?? ''}`
  const iconClass = `img-icontext-btn-${size}`
  const titleClass = `icontext-title icontext-title-${size} ${
    additionalClass ?? ''
  }`
  const subTitleClass = `icontext-title icontext-sub-title ${
    additionalClass ?? ''
  }`

  return (
    <button className={containerClass} {...props}>
      {icon && <img className={iconClass} alt='icon' src={icon} />}
      {title && <span className={titleClass}>{title}</span>}
      {subTitle && <span className={subTitleClass}>{subTitle}</span>}
    </button>
  )
}

export default IconTextButton
