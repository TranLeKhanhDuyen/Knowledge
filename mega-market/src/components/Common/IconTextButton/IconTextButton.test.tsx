import { render, fireEvent } from '@testing-library/react'
import IconTextButton, { IIconTextButtonProps } from '.'

// Mock IconSvg component
jest.mock(
  '@components/Common/IconSvg',
  () =>
    ({ name, className }: { name: string; className?: string }) => (
      <svg data-testid={`icon-${name}`} className={className} />
    )
)

describe('IconTextButton', () => {
  const renderIconTextButton = (props: Partial<IIconTextButtonProps> = {}) => {
    const defaultProps: IIconTextButtonProps = {
      size: 'sm',
      iconPosition: 'left',
      title: 'Default Title',
      subTitle: 'Default SubTitle',
      icon: 'search',
      additionalClass: '',
      onClick: jest.fn()
    }
    return render(<IconTextButton {...defaultProps} {...props} />)
  }

  test('renders with the correct size class', () => {
    const { container } = renderIconTextButton({ size: 'lg' })
    const buttonElement = container.querySelector('button')
    expect(buttonElement).toHaveClass('icontext-btn-container-lg')
  })

  test('renders with the correct icon position class', () => {
    const { container } = renderIconTextButton({ iconPosition: 'right' })
    const buttonElement = container.querySelector('button')
    expect(buttonElement).toHaveClass('icontext-btn-container-right')
  })

  test('renders the correct icon', () => {
    const { getByTestId } = renderIconTextButton({ icon: 'user' })
    expect(getByTestId('icon-user')).toBeInTheDocument()
  })

  test('applies additional class names', () => {
    const { container } = renderIconTextButton({
      additionalClass: 'custom-class'
    })
    const buttonElement = container.querySelector('button')
    expect(buttonElement).toHaveClass('custom-class')
  })

  test('renders the correct title and subtitle', () => {
    const { getByText } = renderIconTextButton({
      title: 'Test Title',
      subTitle: 'Test SubTitle'
    })
    expect(getByText('Test Title')).toBeInTheDocument()
    expect(getByText('Test SubTitle')).toBeInTheDocument()
  })

  test('renders children correctly', () => {
    const { getByText } = renderIconTextButton({
      children: <span>Child Content</span>
    })
    expect(getByText('Child Content')).toBeInTheDocument()
  })

  test('handles the onClick event', () => {
    const onClickMock = jest.fn()
    const { getByRole } = renderIconTextButton({ onClick: onClickMock })
    const buttonElement = getByRole('button')
    fireEvent.click(buttonElement)
    expect(onClickMock).toHaveBeenCalledTimes(1)
  })

  test('matches snapshot for default props', () => {
    const { asFragment } = renderIconTextButton()
    expect(asFragment()).toMatchSnapshot()
  })

  test('matches snapshot for size=lg and iconPosition=right', () => {
    const { asFragment } = renderIconTextButton({
      size: 'lg',
      iconPosition: 'right'
    })
    expect(asFragment()).toMatchSnapshot()
  })

  test('matches snapshot for custom class and no subtitle', () => {
    const { asFragment } = renderIconTextButton({
      additionalClass: 'custom-class',
      subTitle: undefined
    })
    expect(asFragment()).toMatchSnapshot()
  })
})
