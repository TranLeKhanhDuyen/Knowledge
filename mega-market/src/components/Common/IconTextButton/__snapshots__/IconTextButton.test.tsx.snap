// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconTextButton matches snapshot for custom class and no subtitle 1`] = `
<DocumentFragment>
  <button
    class="icontext-btn-container icontext-btn-container-sm  custom-class"
  >
    <svg
      class="img-icontext-btn img-icontext-btn-sm"
      data-testid="icon-search"
    />
    <span
      class="icontext-title icontext-title-sm "
    >
      Default Title
    </span>
  </button>
</DocumentFragment>
`;

exports[`IconTextButton matches snapshot for default props 1`] = `
<DocumentFragment>
  <button
    class="icontext-btn-container icontext-btn-container-sm  "
  >
    <svg
      class="img-icontext-btn img-icontext-btn-sm"
      data-testid="icon-search"
    />
    <span
      class="icontext-title icontext-title-sm "
    >
      Default Title
    </span>
    <span
      class="icontext-title icontext-sub-title "
    >
      Default SubTitle
    </span>
  </button>
</DocumentFragment>
`;

exports[`IconTextButton matches snapshot for size=lg and iconPosition=right 1`] = `
<DocumentFragment>
  <button
    class="icontext-btn-container icontext-btn-container-lg icontext-btn-container-right "
  >
    <svg
      class="img-icontext-btn img-icontext-btn-lg"
      data-testid="icon-search"
    />
    <span
      class="icontext-title icontext-title-lg "
    >
      Default Title
    </span>
    <span
      class="icontext-title icontext-sub-title "
    >
      Default SubTitle
    </span>
  </button>
</DocumentFragment>
`;
